Framework/API/src/LogManager.cpp

```diff
@@ -176,21 +179,74 @@ const Types::Core::DateAndTime LogManager::endTime() const {
     }
   }
 
-  if (hasProperty("proton_charge")) {
-    try {
-      Kernel::TimeSeriesProperty<double> *protonchargelog =
-          dynamic_cast<Kernel::TimeSeriesProperty<double> *>(getProperty("proton_charge"));
-      return DateAndTime(protonchargelog->lastTime());
-    } catch (std::invalid_argument &) { /*Swallow and move on*/
-    }
+  try {
+    return getLastPulseTime();
+  } catch (Exception::NotFoundError &) { /*Swallow and move on*/
+  } catch (std::invalid_argument &) {    /*Swallow and move on*/
   }
 
   throw std::runtime_error("No valid end time has been set for this run.");
 }
 
-bool LogManager::hasStartTime() const { return hasProperty(START_TIME_NAME) || hasProperty("run_start"); }
+bool LogManager::hasStartTime() const {
+  try {
+    auto temp = startTime();
+    UNUSED_ARG(temp);
+  } catch (...) {
+    return false;
+  }
+  return true;
+}
+
+bool LogManager::hasEndTime() const {
+  try {
+    auto temp = endTime();
+    UNUSED_ARG(temp);
+  } catch (...) {
+    return false;
+  }
+  return true;
+}
+
+/** Return the time of the first pulse received, by accessing the run's
+ * sample logs to find the proton_charge.
+ *
+ * NOTE, JZ: Pulse times before 1991 (up to 100) are skipped. This is to avoid
+ * a DAS bug at SNS around Mar 2011 where the first pulse time is Jan 1, 1990.
+ *
+ * @return the time of the first pulse
+ * @throw Exception::NotFoundError if the log is not found; or if it is empty.
+ * @throw invalid_argument if the log is not a double TimeSeriesProperty (should
+ * be impossible)
+ */
+const DateAndTime LogManager::getFirstPulseTime() const {
+  TimeSeriesProperty<double> *log = getTimeSeriesProperty<double>("proton_charge");
+
+  DateAndTime startTime = log->firstTime();
+  DateAndTime reference("1991-01-01T00:00:00");
 
-bool LogManager::hasEndTime() const { return hasProperty(END_TIME_NAME) || hasProperty("run_end"); }
+  int i = 0;
+  // Find the first pulse after 1991
+  while (startTime < reference && i < 100) {
+    i++;
+    startTime = log->nthTime(i);
```

Comment:
`log->nthTime(i)` is somewhat expensive. I think it would be better to cache a copy of `log->timesAsVector()` and go through that. I'm not sure how much it matters when only going through the first 100 values.

Reply:
Done.